### PR TITLE
add controllerManager to manifest

### PR DESCRIPTION
**Pull request by gy95**
**What type of PR is this?**

**What this PR does / why we need it**:
Usage:
Install v1.10.0 kubeedge and controllermanager, by default, we don't install controller, so we should use flag `--set controllerManager.enable=true` to install it, and due to v1.10.0 controller-manager not exist, we use flag  ` --set controllerManager.image.repository=guoyao95/controller-manager` to use our private image
```
_output/local/bin/keadm init --advertise-address=10.0.0.115 --profile version=v1.10.0 --set controllerManager.enable=true --set controllerManager.image.repository=guoyao95/controller-manager
```

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by gy95** [comment]
cc @zhu733756   PTAL :)
I'm not familiar with helm, worried about missing something

File: manifests/charts/cloudcore/templates/deployment_controllermanager.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+kind: Deployment
+metadata:
+  {{- with .Values.controllerManager.labels }}
+  labels: {{- toYaml . | nindent 4 }}
+  {{- end }}
+  name: kubeedge-controller-manager
+  namespace: kubeedge
+spec:
+  selector:
+    {{- with .Values.controllerManager.labels }}
```
Comment (zhu733756) [inline manifests/charts/cloudcore/templates/deployment_controllermanager.yaml:12]: do we need annotations?

File: keadm/cmd/keadm/app/cmd/helm/installer.go [hunk 2]
```diff
@@ -231,7 +232,7 @@ func (cu *KubeCloudHelmInstTool) buildRenderer(baseHelmRoot string) (*Renderer,
 	var subDir string
 	if cu.existsProfile && cu.isInnerProfile() {
 		switch cu.ProfileKey {
-		case VersionProfileKey, IptablesMgrProfileKey:
+		case VersionProfileKey, IptablesMgrProfileKey, ControllerManagerProfileKey:
```
Comment (gy95) [inline keadm/cmd/keadm/app/cmd/helm/installer.go:235]: Is this right? @zhu733756

File: manifests/charts/cloudcore/values.yaml [hunk 3]
```diff
@@ -99,3 +99,32 @@ iptablesManager:
     requests:
       cpu: 100m
       memory: 25Mi
+
+controllerManager:
+  enable: false
+  hostNetWork: true
+  image:
```
Comment (zhu733756) [inline manifests/charts/cloudcore/values.yaml:106]: Is this component needed host network? I'm not familiar with it.

File: keadm/cmd/keadm/app/cmd/helm/installer.go [hunk 4]
```diff
@@ -397,6 +399,10 @@ func (cu *KubeCloudHelmInstTool) handleProfile(profileValue string) error {
 		}
 
 		return fmt.Errorf("the given mode of iptablesmgr %s is not supported, only support internal or external", profileValue)
+
+	case ControllerManagerProfileKey:
+		cu.Sets = append(cu.Sets, fmt.Sprintf("%s=%s", "controllerManager.image.tag", currentVersion))
```
Comment (zhu733756) [inline keadm/cmd/keadm/app/cmd/helm/installer.go:404]: the handling logic of the image tag should be consistent?

**Comment by zhu733756** [comment]
Do we need a svc to post some api to this component?

File: manifests/charts/cloudcore/values.yaml (hunk 3 above)
Comment (gy95) [reply manifests/charts/cloudcore/values.yaml:106]: no, these deployment files are inherit from dir https://github.com/kubeedge/kubeedge/tree/master/build/controllermanager

File: keadm/cmd/keadm/app/cmd/helm/installer.go (hunk 2 above)
Comment (zhu733756) [reply keadm/cmd/keadm/app/cmd/helm/installer.go:235]: > Is this right? @zhu733756

Yes

**Comment by gy95** [comment]
> Do we need a svc to post some api to this component?

maybe not, I don't find svc in dir https://github.com/kubeedge/kubeedge/tree/master/build/controllermanager
WDYT @vincentgoat @Congrool

File: manifests/charts/cloudcore/templates/deployment_controllermanager.yaml (hunk 1 above)
Comment (gy95) [reply manifests/charts/cloudcore/templates/deployment_controllermanager.yaml:15]: added. please help check again :)

File: manifests/charts/cloudcore/values.yaml (hunk 3 above)
Comment (wackxu) [reply manifests/charts/cloudcore/values.yaml:105]: it does not need host network

**Comment by zhu733756** [comment]
lgtm

**Comment by Congrool** [comment]
I'm not sure if we cloud set the image tag through `--set controllerManager.image.tag=v1.10.0-beta.0-264-g315c619`. I tried this , but the image tag for the controller-manager was still `v1.10.0`.

**Comment by gy95** [comment]
maybe this is due to `--profile version=v1.10.0 ` overwrite the controllermanager version, can just remove `--profile version=v1.10.0 ` and try again

**Comment by gy95** [comment]
@Congrool  the second commit fix this bug, it's due to `cu.Sets` contains duplicated string, If we specify `controllerManager.image.tag` in cmd flag, we don't need to contruct it again with `--profile version=v1.10.0`
cc @zhu733756   please help review the second commit whether is OK. :)

**Comment by zhu733756** [comment]
thanks @gy95 

/lgtm

**Comment by Congrool** [comment]
lgtm

File: keadm/cmd/keadm/app/cmd/helm/installer.go [hunk 5]
```diff
@@ -381,8 +383,15 @@ func (cu *KubeCloudHelmInstTool) handleProfile(profileValue string) error {
 			profileValue = "v" + profileValue
 		}
 
-		cu.Sets = append(cu.Sets, fmt.Sprintf("%s=%s", "cloudCore.image.tag", profileValue))
-		cu.Sets = append(cu.Sets, fmt.Sprintf("%s=%s", "iptablesManager.image.tag", profileValue))
+		if !SetsContainSubstring(cu.Sets, "cloudCore.image.tag") {
```
Comment (wackxu) [inline keadm/cmd/keadm/app/cmd/helm/installer.go:386]: the behavior is changed vs before, it is right `!SetsContainSubstring`?

File: manifests/charts/cloudcore/templates/rbac_controllermanager.yaml [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+rules:
+  - apiGroups: ["apps.kubeedge.io"]
+    resources: ["nodegroups", "nodegroups/status"]
+    verbs: ["*"]
+  - apiGroups: [""]
+    resources: ["nodes"]
+    verbs: ["list", "watch", "patch", "get"]
+  - apiGroups: [""]
+    resources: ["pods"]
+    verbs: ["list", "watch", "get", "delete"]
```
Comment (wackxu) [inline manifests/charts/cloudcore/templates/rbac_controllermanager.yaml:45]: do we need delete pod?

Comment (Congrool) [reply manifests/charts/cloudcore/templates/rbac_controllermanager.yaml:45]: Oh, actually it does not delete the pod. I'll revise it.

File: keadm/cmd/keadm/app/cmd/helm/installer.go (hunk 5 above)
Comment (gy95) [reply keadm/cmd/keadm/app/cmd/helm/installer.go:386]: yes, without this, cu.Sets will contain multiple duplicated `cloudCore.image.tag`, one is here `--profile version=v1.10.0`, and will also contain one if we specify in cmd line `--set cloudCore.image.tag=v1.10.0`

Comment (wackxu) [reply keadm/cmd/keadm/app/cmd/helm/installer.go:386]: What is the priority of these two parameters？if user build their own cloudcore image with a different tag, how to  achieve this?

File: manifests/charts/cloudcore/templates/deployment_controllermanager.yaml [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+        - name: controller-manager
+          image: {{ .Values.controllerManager.image.repository }}:{{ .Values.controllerManager.image.tag }}
+          imagePullPolicy: {{ .Values.controllerManager.image.pullPolicy }}
+      restartPolicy: Always
+      serviceAccountName: controller-manager
+      {{- with .Values.controllerManager.affinity }}
+      affinity: {{ toYaml . | nindent 8 }}
+      {{- end }}
+      {{- with .Values.controllerManager.tolerations }}
+      tolerations: {{ toYaml . | nindent 8 }}
```
Comment (wackxu) [inline manifests/charts/cloudcore/templates/deployment_controllermanager.yaml:34]: suggest add imagePullSecrets to allow user config it

File: keadm/cmd/keadm/app/cmd/helm/installer.go (hunk 5 above)
Comment (gy95) [reply keadm/cmd/keadm/app/cmd/helm/installer.go:386]: --set has a higher priority, If we don't use `--set` to set a image tag, we can use `--profile version=v1.10.0` to specify one uniform image tag(this tag will apply for all images, including cloudcore, iptables-manager, and controller-manager)

for example, if we want to use v1.10.0 cloudcore image and v1.11.0 controller-manager, we can use the below commd
```
_output/local/bin/keadm init --advertise-address=10.0.0.115 --profile version=v1.10.0 --set controllerManager.enable=true --set controllerManager.image.tag=v1.11.0
```

Comment (wackxu) [reply keadm/cmd/keadm/app/cmd/helm/installer.go:386]: got it, thanks for your explain

File: manifests/charts/cloudcore/templates/rbac_controllermanager.yaml (hunk 6 above)
Comment (Congrool) [reply manifests/charts/cloudcore/templates/rbac_controllermanager.yaml:45]: Sorry for my mistake, the authorization for deleting pod is needed for nodegroup controller. I'll explain the scenario:

Considering that we have an eapp deployed in the cluster which having pods running on `node-a` and `node-b`, and `node-a` `node-b` are belonging to `nodegroup-hangzhou`, when we remove `node-a` from `nodegroup-hangzhou`, we should evict the pods of eapp running on `node-a`, and then create new instances on `node-b` or other nodes in `nodegroup-hangzhou`.   The eviction needs to delete the pod.

Comment (wackxu) [reply manifests/charts/cloudcore/templates/rbac_controllermanager.yaml:45]: I think we should not do this is nodegroup controller, In edgeapplication, we manage resouces such as deployment， statefulset and so on, we donot manage pods directly. when user delete a pod,  he should do this carefully as below
https://kubernetes.io/zh-cn/docs/tasks/administer-cluster/safely-drain-node/
when he drain the node, all pod on this node will be evict. and the k8s kube-controller will watch this and then create new pod for the deployment or statefulset.

Comment (wackxu) [reply manifests/charts/cloudcore/templates/rbac_controllermanager.yaml:45]: there will be conflict when we mange pod vs k8s controller manager

File: manifests/charts/cloudcore/templates/deployment_controllermanager.yaml (hunk 7 above)
Comment (wackxu) [reply manifests/charts/cloudcore/templates/deployment_controllermanager.yaml:34]: I think we should also add priority for all manager component

File: manifests/charts/cloudcore/templates/rbac_controllermanager.yaml (hunk 6 above)
Comment (Congrool) [reply manifests/charts/cloudcore/templates/rbac_controllermanager.yaml:45]: You're right when deleting a node from cluster. But other scenarios we should take into consideration. Let's continue with the above example:  
Situation 1: `node-a` is not deleted from cluster, but just be an orphan node, which means it does not belong to any nodegroup.
Situation 2: `node-a` is moved from `nodegroup-hangzhou` to `nodegroup-beijing`.

The original idea of eviciting pods on node is to handle the above 2 situations, but the situation you mentioned above seems also be included. I need to figure out if there's competation between nodegroup controller and k8s controller manager when deleting nodes.

**Comment by gy95** [comment]
cc @fisherxu

**Comment by gy95** [comment]
cc @fisherxu  PTAL

**Review by fisherxu — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**